src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.cpp

```diff
@@ -0,0 +1,910 @@
+  out << "cbmc_loop_id: " << cbmc_loop_id << "\n";
+  out << "local: {";
+  for(auto &id : local)
+  {
+    out << id << ", ";
+  }
+  out << "}\n";
+
+  out << "tracked: {";
+  for(auto &id : tracked)
+  {
+    out << id << ", ";
+  }
+  out << "}\n";
+
+  out << "write_set: " << format(write_set_var) << "\n";
+
+  out << "addr_of_write_set: " << format(addr_of_write_set_var) << "\n";
+
+  out << "assigns: {";
+  for(auto &expr : assigns)
+  {
+    out << format(expr) << ", ";
+  }
+  out << "}\n";
+
+  out << "invariant: " << format(invariant) << "\n";
+
+  out << "decreases: {";
+  for(auto &expr : decreases)
+  {
+    out << format(expr) << ", ";
+  }
+  out << "}\n";
+
+  out << "inner loops: {"
+      << "\n";
+  for(auto &id : inner_loops)
+  {
+    out << id << ", ";
+  }
+  out << "}"
+      << "\n";
+
+  out << "outer loops: {"
+      << "\n";
+  for(auto &id : outer_loops)
+  {
+    out << id << ", ";
+  }
+  out << "}"
+      << "\n";
+}
+
+optionalt<goto_programt::targett>
+dfcc_loop_infot::find_head(goto_programt &goto_program) const
+{
+  for(auto target = goto_program.instructions.begin();
+      target != goto_program.instructions.end();
+      target++)
+  {
+    if(dfcc_is_loop_head(target) && dfcc_has_loop_id(target, loop_id))
+    {
+      return target;
+    }
+  }
+  return {};
+}
+
+optionalt<goto_programt::targett>
+dfcc_loop_infot::find_latch(goto_programt &goto_program) const
+{
+  optionalt<goto_programt::targett> result = nullopt;
+  for(auto target = goto_program.instructions.begin();
+      target != goto_program.instructions.end();
+      target++)
+  {
+    // go until the end because we want to find the very last occurrence
+    if(dfcc_is_loop_latch(target) && dfcc_has_loop_id(target, loop_id))
+    {
+      result = target;
+    }
+  }
+  return result;
+}
+
+/// Check if there exists some loop with contracts while some of its
+/// nested loops contains no contract.
+/// Return the loop head if found a loop without contract nested
+/// in a loop with a contract.
+/// Return None otherwise.
+static optionalt<goto_programt::targett> check_inner_loops_have_contracts(
+  const dfcc_loop_nesting_grapht &loop_nesting_graph)
+{
+  std::list<size_t> must_have_contracts;
+  for(std::size_t idx = 0; idx < loop_nesting_graph.size(); idx++)
+  {
+    // keep track of which loops have contracts
+    const auto &node = loop_nesting_graph[idx];
+    if(has_contract(node.latch))
+    {
+      for(const auto pred_idx : loop_nesting_graph.get_predecessors(idx))
+        must_have_contracts.push_back(pred_idx);
+    }
+  }
+
+  while(!must_have_contracts.empty())
+  {
+    const auto idx = must_have_contracts.front();
+    must_have_contracts.pop_front();
+    const auto &node = loop_nesting_graph[idx];
+    if(!has_contract(node.latch))
+    {
+      return node.head;
+    }
+
+    for(const auto pred_idx : loop_nesting_graph.get_predecessors(idx))
+      must_have_contracts.push_back(pred_idx);
+  }
+  return {};
```

Comment:
We allow that a inner loop has contract while its outer loop does not. So if we starting from top-level loops, it can be that the top_level loop `A` does not have loops contracts. But there are inner loop `B` with contracts and `C` without contracts where `C` nested in `B` and `B` nested in `A`.

Reply:
Yes we start a traversal from each loop that has a contract on purpose, at the cost of visiting some loops more than once. As @qinheping explained, we allow programs where a top level loop A has no contract, but has an inner loop B with a contract, then any loop C nested in B must also have a contract.

I can rewrite this with a recursive function that starts from top level loops only but this is really apples-to-apples

Reply:
replaced by a recursive function